SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -39,180 +39,175 @@ void SpatialEventTracer::TraceCallback(void* UserData, const Trace_Item* Item)
+
+	// Setup Event Tracing Sampling
+	Trace_SamplingParameters SamplingParameters;
+	SamplingParameters.sampling_mode = Trace_SamplingMode::TRACE_SAMPLING_MODE_PROBABILISTIC;
+
+	TArray<Trace_SpanSamplingProbability> SpanSamplingProbabilities;
+	TArray<TStringConversion<TStringConvert<TCHAR, ANSICHAR>, 128>> AnsiTypes;
+
+	for (const auto& Pair : Settings->EventSamplingModeOverrides)
 	{
-		if (Settings->bEventTracingEnabled)
-		{
-			MaxFileSize = Settings->MaxEventTracingFileSizeBytes;
-			Enable(WorkerId);
-		}
+		int32 Index = AnsiTypes.Add(StringCast<ANSICHAR>(*Pair.Key.ToString()));
+		SpanSamplingProbabilities.Add({ AnsiTypes[Index].Get(), Pair.Value });
 	}
-}
 
-SpatialEventTracer::~SpatialEventTracer()
-{
-	if (IsEnabled())
+	SamplingParameters.probabilistic_parameters.default_probability = Settings->SamplingProbability;
+	SamplingParameters.probabilistic_parameters.probability_count = SpanSamplingProbabilities.Num();
+	SamplingParameters.probabilistic_parameters.probabilities = SpanSamplingProbabilities.GetData();
+
+	Trace_EventTracer_SetSampler(EventTracer, &SamplingParameters);
+
+	UE_LOG(LogSpatialEventTracer, Log, TEXT("Spatial event tracing enabled."));
+
+	// Open a local file
+	FString EventTracePath = FPaths::Combine(FPaths::ProjectSavedDir(), TEXT("EventTracing"));
+	FString AbsLogPath;
+	if (FParse::Value(FCommandLine::Get(), TEXT("eventLogPath="), AbsLogPath, false))
 	{
-		UE_LOG(LogSpatialEventTracer, Log, TEXT("Spatial event tracing disabled."));
-		Trace_EventTracer_Disable(EventTracer);
-		Trace_EventTracer_Destroy(EventTracer);
+		EventTracePath = FPaths::GetPath(AbsLogPath);
 	}
-}
 
-FString SpatialEventTracer::SpanIdToString(const Trace_SpanId& SpanId)
-{
-	FString HexStr;
-	for (int i = 0; i < TraceSpanIdLength; i++)
+	FolderPath = EventTracePath;
+	const FString FullFileName = FString::Printf(TEXT("EventTrace_%s_%s.trace"), *WorkerId, *FDateTime::Now().ToString());
+	const FString FilePath = FPaths::Combine(FolderPath, FullFileName);
+
+	IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+	if (PlatformFile.CreateDirectoryTree(*FolderPath))
 	{
-		HexStr += FString::Printf(TEXT("%02x"), SpanId.data[i]);
+		UE_LOG(LogSpatialEventTracer, Log, TEXT("Capturing trace to %s."), *FilePath);
+		Stream.Reset(Io_CreateFileStream(TCHAR_TO_ANSI(*FilePath), Io_OpenMode::IO_OPEN_MODE_WRITE));
 	}
-	return HexStr;
 }
 
-FUserSpanId SpatialEventTracer::SpanIdToUserSpanId(const Trace_SpanId& SpanId)
+SpatialEventTracer::~SpatialEventTracer()
+{
+	UE_LOG(LogSpatialEventTracer, Log, TEXT("Spatial event tracing disabled."));
+	Trace_EventTracer_Destroy(EventTracer);
+}
+
+FUserSpanId SpatialEventTracer::GDKSpanIdToUserSpanId(const FSpatialGDKSpanId& SpanId)
 {
 	FUserSpanId UserSpanId;
-	UserSpanId.Data.SetNum(TraceSpanIdLength);
-	const int32 Size = TraceSpanIdLength * sizeof(uint8);
-	FMemory::Memcpy(UserSpanId.Data.GetData(), SpanId.data, Size);
+	UserSpanId.Data.SetNum(TRACE_SPAN_ID_SIZE_BYTES);
```

Comment:
Where is this const coming from?

Reply:
Are you referring to "GetConstId"? If so, it's just a function on the FSpatialGDKSpanId the gets a const pointer to it's wrapped up internal Id.

Reply:
Ah nah I meant `TRACE_SPAN_ID_SIZE_BYTES` it's not obvious without a namespace